core/blockchain.go

```diff
@@ -1212,22 +1216,27 @@ func (bc *BlockChain) WriteBlockWithState(block *types.Block, receipts []*types.
 
 	//// Cross-links
 	if len(header.CrossLinks()) > 0 {
+		header.Logger(utils.Logger()).Debug().Msg("[insertChain/crosslinks] writing crosslinks into blockchain...")
 		crossLinks := &types.CrossLinks{}
 		err = rlp.DecodeBytes(header.CrossLinks(), crossLinks)
 		if err != nil {
-			header.Logger(utils.Logger()).Warn().Err(err).Msg("[insertChain] cannot parse cross links")
+			header.Logger(utils.Logger()).Warn().Err(err).Msg("[insertChain/crosslinks] cannot parse cross links")
 			return NonStatTy, err
 		}
 		if !crossLinks.IsSorted() {
-			header.Logger(utils.Logger()).Warn().Err(err).Msg("[insertChain] cross links are not sorted")
+			header.Logger(utils.Logger()).Warn().Err(err).Msg("[insertChain/crosslinks] cross links are not sorted")
 			return NonStatTy, errors.New("proposed cross links are not sorted")
 		}
 		for _, crossLink := range *crossLinks {
-			if err := bc.WriteCrossLinks(types.CrossLinks{crossLink}, false); err == nil {
-				utils.Logger().Info().Uint64("blockNum", crossLink.BlockNum().Uint64()).Uint32("shardID", crossLink.ShardID()).Msg("[InsertChain] Cross Link Added to Beaconchain")
+			if err := bc.WriteCrossLinks(types.CrossLinks{crossLink}); err == nil {
+				utils.Logger().Info().Uint64("blockNum", crossLink.BlockNum()).Uint32("shardID", crossLink.ShardID()).Msg("[insertChain/crosslinks] Cross Link Added to Beaconchain")
 			}
-			bc.DeleteCrossLinks(types.CrossLinks{crossLink}, true)
-			bc.WriteShardLastCrossLink(crossLink.ShardID(), crossLink)
+			bc.UpdateShardLastCrossLink(crossLink.ShardID(), crossLink)
+		}
+		//clean local database cache after crosslink inserted into blockchain
+		err = bc.DeletePendingCrossLinks()
```

Comment:
why delete here? there could be pending cross links that's not proposed in the block.

Reply:
checking and deleting only existing cross links